### PR TITLE
raidemulator: Convert EventBus to typescript

### PR DESCRIPTION
**Pull request by valarnin**
Lots of weird issues with this one.

### CODE REVIEW

File: ui/raidboss/emulator/EventBus.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
+type CallbackFunction = () => void;
+type EventMapEntry = {
+  event: string;
+  // Disable ban-types due to scope being any object type
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  scope: object;
+  callback: CallbackFunction;
+};
```
Comment (quisquous) [inline ui/raidboss/emulator/EventBus.ts:8]:
```suggestion
// Disable ban-types due to scope being any object type
// eslint-disable-next-line @typescript-eslint/ban-types
type Scope = object;

type EventMapEntry = {
  event: string;
  scope: Scope;
  callback: CallbackFunction;
};
```

I think better to not repeat the disable everywhere and just keep the "problematic" type in one place.

File: ui/raidboss/emulator/EventBus.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+   * @param event The event(s) to subscribe to, space separated
+   * @param callback The callback to invoke
+   * @param scope Optional. The scope to apply the function against
+   * @returns The callbacks registered to the event(s)
+   */
+  on(event: string,
+      callback: CallbackFunction | undefined = undefined,
+      // Disable ban-types due to scope being any object type
+      // eslint-disable-next-line @typescript-eslint/ban-types
+      scope: object | undefined = undefined,
```
Comment (quisquous) [inline ui/raidboss/emulator/EventBus.ts:30]:
```suggestion
      callback?: CallbackFunction,
      scope?: Scope,
```

Along with the above suggestion for `Scope`, and I think `| undefined = undefined` can just be `?`.

File: ui/raidboss/emulator/EventBus.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+   * @param event The event to dispatch
+   * @param eventArguments The event arguments to pass to listeners
+   * @returns A promise that can be await'd or ignored
+   */
+  // Disable no-expicit-any and explicit-module-boundary-types for eventArguments since it's
+  // deliberately any-type/generic. Also disable max-len for the eslint syntax line because it's
+  // too long.
+  // eslint-disable-next-line max-len
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
+  async dispatch(event: string, ...eventArguments: any): Promise<undefined> {
```
Comment (quisquous) [inline ui/raidboss/emulator/EventBus.ts:56]:
```suggestion
  async dispatch(event: string, ...eventArguments: any): Promise<void> {
```

File: ui/raidboss/emulator/EventBus.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+   * Dispatch an event to any subscribers
+   *
+   * @param event The event to dispatch
+   * @param eventArguments The event arguments to pass to listeners
+   * @returns A promise that can be await'd or ignored
+   */
+  // Disable no-expicit-any and explicit-module-boundary-types for eventArguments since it's
+  // deliberately any-type/generic. Also disable max-len for the eslint syntax line because it's
+  // too long.
+  // eslint-disable-next-line max-len
```
Comment (quisquous) [inline ui/raidboss/emulator/EventBus.ts:54]: :rofl:

Yeah, I think a generic event bus like this is going to necessarily be generic and not be able to have strict typing.  I'm not sure there's anything that can really be done while keeping this anywhere near the same form it is now.

File: ui/raidboss/emulator/EventBus.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+   * @param callback The callback to invoke
+   * @param scope Optional. The scope to apply the function against
+   * @returns The callbacks registered to the event(s)
+   */
+  on(event: string,
+      callback: CallbackFunction | undefined = undefined,
+      // Disable ban-types due to scope being any object type
+      // eslint-disable-next-line @typescript-eslint/ban-types
+      scope: object | undefined = undefined,
+  ): EventMapEntry[] | undefined {
```
Comment (quisquous) [inline ui/raidboss/emulator/EventBus.ts:31]:
```suggestion
  ): EventMapEntry[] {
```

**Comment by valarnin** [comment]
Took the review suggestions and added a bit of extra cleanup on top of them.

File: ui/raidboss/emulator/EventBus.ts [hunk 6]
```diff
@@ -48,12 +48,7 @@ export default class EventBus {
    * @param eventArguments The event arguments to pass to listeners
    * @returns A promise that can be await'd or ignored
    */
-  // Disable no-expicit-any and explicit-module-boundary-types for eventArguments since it's
-  // deliberately any-type/generic. Also disable max-len for the eslint syntax line because it's
-  // too long.
-  // eslint-disable-next-line max-len
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
-  async dispatch(event: string, ...eventArguments: any): Promise<undefined> {
+  async dispatch(event: string, ...eventArguments: Param): Promise<void> {
```
Comment (quisquous) [inline ui/raidboss/emulator/EventBus.ts:51]: :+1:

**Pull request merged**